src/Store/Psr14Store.php

```diff
@@ -0,0 +1,159 @@
+    ) {
+        [$sessionPrefix] = Toolkit::filter([$sessionPrefix])->string()->trim();
+
+        Toolkit::assert([
+            [$sessionPrefix, \Auth0\SDK\Exception\ArgumentException::missing('sessionPrefix')],
+        ])->isString();
+
+        $this->configuration = $configuration;
+        $this->sessionPrefix = $sessionPrefix ?? 'auth0';
+    }
+
+    /**
+     * Dispatch event to notify that the class is being destructed.
+     */
+    public function __destruct()
+    {
+        $this->configuration->eventDispatcher()->dispatch(new Destruct($this));
+    }
+
+    /**
+     * Dispatch event to toggle state deferrance.
+     *
+     * @param bool $deferring Whether to defer persisting the storage state.
+     */
+    public function defer(
+        bool $deferring
+    ): void {
+        $this->boot();
+        $this->configuration->eventDispatcher()->dispatch(new Defer($this, $deferring));
+    }
+
+    /**
+     * Dispatch event to set the value of a key-value pair.
+     *
+     * @param string $key   Session key to set.
+     * @param mixed  $value Value to use.
+     */
+    public function set(
+        string $key,
+        $value
+    ): void {
+        $this->boot();
+        $this->configuration->eventDispatcher()->dispatch(new Set($this, $key, $value));
+    }
+
+    /**
+     * Dispatch event to retrieve the value of a key-value pair.
+     *
+     * @param string $key     Session key to query.
+     * @param mixed  $default Default to return if nothing was found.
+     *
+     * @return mixed
+     */
+    public function get(
+        string $key,
+        $default = null
+    ) {
+        $this->boot();
+
+        /**
+         * @var Get $event
+         */
+        $event = $this->configuration->eventDispatcher()->dispatch(new Get($this, $key));
+        $value = $event->getValue();
+
+        if ($value === null) {
+            return $default;
+        }
+
+        return $value;
+    }
+
+    /**
+     * Dispatch event to clear all key-value pairs.
+     */
+    public function purge(): void
+    {
+        $this->boot();
+        $this->configuration->eventDispatcher()->dispatch(new Clear($this));
+    }
+
+    /**
+     * Dispatch event to delete key-value pair.
+     *
+     * @param string $key Session key to delete.
+     */
+    public function delete(
+        string $key
+    ): void {
+        $this->boot();
+        $this->configuration->eventDispatcher()->dispatch(new Delete($this, $key));
+    }
+
+    /**
+     * Dispatch event to alert that a session should be prepared for an incoming request.
+     */
+    private function boot(): void
+    {
+        if (! $this->booted) {
+            $event = $this->configuration->eventDispatcher()->dispatch(new Boot($this, $this->sessionPrefix));
+
+            /**
+             * @var Boot $event
+             */
+
+            $this->sessionPrefix = $event->getPrefix();
+            $this->booted = true;
+        }
+
+        return;
```

Comment:
Why is this a separate method, instead of just being doing in the constructor of the class?

Reply:
Hey @shadowhand 👋

In part this is to avoid problems in downstream frameworks that support lazy loading of classes. For customers that rely on that, I try to avoid manipulating state data within constructors as often as I can.

Also to avoid creating scenarios where downstream apps might unnecessarily open database connections or session streams at SDK initialization in response to the Boot event, but session data may ultimately never even get touched. Dispatching the Boot event and state right before it's necessary is just efficient.